AndroidSDKTests/src/test/java/com/leanplum/internal/RequestTest.java

```diff
@@ -67,6 +71,39 @@ public void setUp() {
     Leanplum.setApplicationContext(context);
   }
 
+  /**
+   * Test that read writes happened sequentially when calling sendNow().
+   */
+  @Test
+  public void shouldWriteRequestAndSendInSequence() throws InterruptedException {
+    // Given a request.
+    Map<String, Object> params = new HashMap<>();
+    params.put("data1", "value1");
+    params.put("data2", "value2");
+    final ThreadRequestSequence threadRequestSequence = new ThreadRequestSequence();
+    Request request = new Request(POST, Constants.Methods.START, params, threadRequestSequence);
+    request.setAppId("fskadfshdbfa", "wee5w4waer422323");
+
+    new Thread(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          Thread.sleep(100);
+        } catch (InterruptedException e) {
+          throw new RuntimeException(e);
+        }
+        threadRequestSequence.writeSemaphore.release(1);
+      }
+    }).start();
+
+    // When the request is sent.
+    request.sendIfConnected();
+
+    threadRequestSequence.testThreadSemaphore.tryAcquire(5000, TimeUnit.MILLISECONDS);
+
+    // When the request is sent.
```

Comment:
update comment, suggestion

// Then the request is written to the local db first, and then read and sent.